File: mobile/app-flutter/lib/services/maviance_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class MavianceService {
  static const String baseUrl = 'http://localhost:3000/api';
  static const String apiKey = 'votre_cle_api_secrete'; // Même clé que dans le backend

  static Map<String, String> get headers => {
    'Content-Type': 'application/json',
    'x-api-key': apiKey,
  };

  // Test de ping
  static Future<Map<String, dynamic>> ping() async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/ping'),
        headers: headers,
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        return {
          'success': false,
          'error': 'HTTP ${response.statusCode}: ${response.body}'
        };
      }
    } catch (e) {
      return {
        'success': false,
        'error': 'Erreur réseau: $e'
      };
    }
  }

  // Recharge de crédit
  static Future<Map<String, dynamic>> rechargeCredit({
    required String phoneNumber,
    required double amount,
    required String payItemId,
    required Map<String, String> customerInfo,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/recharge'),
        headers: headers,
        body: jsonEncode({
          'phoneNumber': phoneNumber,
          'amount': amount,
          'payItemId': payItemId,
          'customerInfo': customerInfo,
        }),
      );

      return jsonDecode(response.body);
    } catch (e) {
      return {
        'success': false,
        'error': 'Erreur réseau: $e'
      };
    }
  }

  // Achat de forfait
  static Future<Map<String, dynamic>> purchaseVoucher({
    required String phoneNumber,
    required String payItemId,
    required Map<String, String> customerInfo,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/voucher'),
        headers: headers,
        body: jsonEncode({
          'phoneNumber': phoneNumber,
          'payItemId': payItemId,
          'customerInfo': customerInfo,
        }),
      );

      return jsonDecode(response.body);
    } catch (e) {
      return {
        'success': false,
        'error': 'Erreur réseau: $e'
      };
    }
  }

  // Dépôt d'argent
  static Future<Map<String, dynamic>> depositMoney({
    required double amount,
    required String payItemId,
    required Map<String, String> customerInfo,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/deposit'),
        headers: headers,
        body: jsonEncode({
          'amount': amount,
          'payItemId': payItemId,
          'customerInfo': customerInfo,
        }),
      );

      return jsonDecode(response.body);
    } catch (e) {
      return {
        'success': false,
        'error': 'Erreur réseau: $e'
      };
    }
  }

  // Retrait d'argent
  static Future<Map<String, dynamic>> withdrawMoney({
    required double amount,
    required String payItemId,
    required Map<String, String> customerInfo,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/withdraw'),
        headers: headers,
        body: jsonEncode({
          'amount': amount,
          'payItemId': payItemId,
          'customerInfo': customerInfo,
        }),
      );

      return jsonDecode(response.body);
    } catch (e) {
      return {
        'success': false,
        'error': 'Erreur réseau: $e'
      };
    }
  }

  // Vérifier une transaction
  static Future<Map<String, dynamic>> verifyTransaction(String transactionId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/verify/$transactionId'),
        headers: headers,
      );

      return jsonDecode(response.body);
    } catch (e) {
      return {
        'success': false,
        'error': 'Erreur réseau: $e'
      };
    }
  }

  // Récupérer les services disponibles
  static Future<Map<String, dynamic>> getServices() async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/services'),
        headers: headers,
      );

      return jsonDecode(response.body);
    } catch (e) {
      return {
        'success': false,
        'error': 'Erreur réseau: $e'
      };
    }
  }

  // Récupérer les produits TOPUP pour un service
  static Future<Map<String, dynamic>> getTopupProducts(String serviceId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/topup/$serviceId'),
        headers: headers,
      );

      return jsonDecode(response.body);
    } catch (e) {
      return {
        'success': false,
        'error': 'Erreur réseau: $e'
      };
    }
  }

  // Récupérer les forfaits pour un service
  static Future<Map<String, dynamic>> getVoucherProducts(String serviceId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/voucher/$serviceId'),
        headers: headers,
      );

      return jsonDecode(response.body);
    } catch (e) {
      return {
        'success': false,
        'error': 'Erreur réseau: $e'
      };
    }
  }

  // Méthodes utilitaires pour générer les données client
  static Map<String, String> generateCustomerInfo({
    required String phone,
    required String email,
    required String name,
    String address = 'Yaoundé, Cameroun',
  }) {
    return {
      'phone': phone,
      'email': email,
      'name': name,
      'address': address,
    };
  }
}

File: backend/api-legacy/maviance_service.js
const axios = require('axios');
const crypto = require('crypto-js');
require('dotenv').config();

class MavianceService {
  constructor() {
    this.baseURL = process.env.S3P_URL;
    this.token = process.env.S3P_KEY;
    this.secret = process.env.S3P_SECRET;
  }

  // Génère la signature d'authentification HMAC-SHA1
  generateAuthHeader(method, url, params = {}) {
    const timestamp = Date.now();
    const nonce = Date.now();
    
    const s3pParams = {
      s3pAuth_nonce: nonce,
      s3pAuth_timestamp: timestamp,
      s3pAuth_signature_method: 'HMAC-SHA1',
      s3pAuth_token: this.token
    };

    const allParams = { ...params, ...s3pParams };
    
    // Nettoie les paramètres (trim les strings)
    Object.keys(allParams).forEach(key => {
      if (typeof allParams[key] === 'string') {
        allParams[key] = allParams[key].trim();
      }
    });

    // Trie les paramètres par ordre alphabétique
    const sortedParams = Object.keys(allParams)
      .sort()
      .reduce((result, key) => {
        result[key] = allParams[key];
        return result;
      }, {});

    // Crée la chaîne de paramètres
    const parameterString = Object.keys(sortedParams)
      .map(key => `${key}=${sortedParams[key]}`)
      .join('&');

    // Crée la base string pour la signature
    const baseString = `${method}&${encodeURIComponent(url)}&${encodeURIComponent(parameterString)}`;
    
    // Génère la signature HMAC-SHA1
    const signature = crypto.HmacSHA1(baseString, this.secret);
    const encodedSignature = crypto.enc.Base64.stringify(signature);

    // Construit l'en-tête d'autorisation
    const authHeader = `s3pAuth s3pAuth_timestamp="${timestamp}", s3pAuth_signature="${encodedSignature}", s3pAuth_nonce="${nonce}", s3pAuth_signature_method="HMAC-SHA1", s3pAuth_token="${this.token}"`;

    return authHeader;
  }

  // Test de ping pour vérifier la connexion
  async ping() {
    try {
      const url = `${this.baseURL}/ping`;
      const authHeader = this.generateAuthHeader('GET', url);
      
      const response = await axios.get(url, {
        headers: {
          'Authorization': authHeader
        }
      });

      return {
        success: true,
        data: response.data
      };
    } catch (error) {
      console.error('Ping failed:', error.response?.data || error.message);
      return {
        success: false,
        error: error.response?.data || error.message
      };
    }
  }

  // Récupère les services disponibles
  async getServices() {
    try {
      const url = `${this.baseURL}/service`;
      const authHeader = this.generateAuthHeader('GET', url);
      
      const response = await axios.get(url, {
        headers: {
          'Authorization': authHeader,
          'x-api-version': '3.0.0'
        }
      });

      return {
        success: true,
        data: response.data
      };
    } catch (error) {
      console.error('Get services failed:', error.response?.data || error.message);
      return {
        success: false,
        error: error.response?.data || error.message
      };
    }
  }

  // Récupère les produits TOPUP (recharge de crédit)
  async getTopupProducts(serviceId) {
    try {
      const url = `${this.baseURL}/topup`;
      const params = { serviceid: serviceId };
      const authHeader = this.generateAuthHeader('GET', url, params);
      
      const response = await axios.get(url, {
        headers: {
          'Authorization': authHeader
        },
        params: params
      });

      return {
        success: true,
        data: response.data
      };
    } catch (error) {
      console.error('Get topup products failed:', error.response?.data || error.message);
      return {
        success: false,
        error: error.response?.data || error.message
      };
    }
  }

  // Récupère les produits VOUCHER (forfaits)
  async getVoucherProducts(serviceId) {
    try {
      const url = `${this.baseURL}/voucher`;
      const params = { serviceid: serviceId };
      const authHeader = this.generateAuthHeader('GET', url, params);
      
      const response = await axios.get(url, {
        headers: {
          'Authorization': authHeader
        },
        params: params
      });

      return {
        success: true,
        data: response.data
      };
    } catch (error) {
      console.error('Get voucher products failed:', error.response?.data || error.message);
      return {
        success: false,
        error: error.response?.data || error.message
      };
    }
  }

  // Crée un devis (quote)
  async createQuote(payItemId, amount) {
    try {
      const url = `${this.baseURL}/quotestd`;
      const body = {
        payItemId: payItemId,
        amount: amount
      };
      
      const authHeader = this.generateAuthHeader('POST', url, body);
      
      const response = await axios.post(url, body, {
        headers: {
          'Authorization': authHeader,
          'Content-Type': 'application/json'
        }
      });

      return {
        success: true,
        data: response.data
      };
    } catch (error) {
      console.error('Create quote failed:', error.response?.data || error.message);
      return {
        success: false,
        error: error.response?.data || error.message
      };
    }
  }

  // Exécute la collecte (paiement)
  async collectPayment(quoteId, customerData, transactionId) {
    try {
      const url = `${this.baseURL}/collectstd`;
      const body = {
        quoteId: quoteId,
        customerPhonenumber: customerData.phone,
        customerEmailaddress: customerData.email,
        customerName: customerData.name,
        customerAddress: customerData.address,
        serviceNumber: customerData.serviceNumber,
        trid: transactionId
      };
      
      const authHeader = this.generateAuthHeader('POST', url, body);
      
      const response = await axios.post(url, body, {
        headers: {
          'Authorization': authHeader,
          'Content-Type': 'application/json'
        }
      });

      return {
        success: true,
        data: response.data
      };
    } catch (error) {
      console.error('Collect payment failed:', error.response?.data || error.message);
      return {
        success: false,
        error: error.response?.data || error.message
      };
    }
  }

  // Vérifie le statut d'une transaction
  async verifyTransaction(transactionId) {
    try {
      const url = `${this.baseURL}/verifytx`;
      const params = { trid: transactionId };
      const authHeader = this.generateAuthHeader('GET', url, params);
      
      const response = await axios.get(url, {
        headers: {
          'Authorization': authHeader
        },
        params: params
      });

      return {
        success: true,
        data: response.data
      };
    } catch (error) {
      console.error('Verify transaction failed:', error.response?.data || error.message);
      return {
        success: false,
        error: error.response?.data || error.message
      };
    }
  }

  // Recharge de crédit (TOPUP)
  async rechargeCredit(phoneNumber, amount, payItemId, customerInfo) {
    try {
      // Étape 1: Créer un devis
      const quoteResult = await this.createQuote(payItemId, amount);
      if (!quoteResult.success) {
        return quoteResult;
      }

      const quoteId = quoteResult.data.quoteId;
      const transactionId = `TR${Date.now()}`;

      // Étape 2: Exécuter le paiement
      const customerData = {
        phone: customerInfo.phone,
        email: customerInfo.email,
        name: customerInfo.name,
        address: customerInfo.address,
        serviceNumber: phoneNumber
      };

      const collectResult = await this.collectPayment(quoteId, customerData, transactionId);
      if (!collectResult.success) {
        return collectResult;
      }

      // Étape 3: Vérifier la transaction (optionnel)
      const verifyResult = await this.verifyTransaction(transactionId);

      return {
        success: true,
        data: {
          transactionId: transactionId,
          quoteId: quoteId,
          collectResult: collectResult.data,
          verifyResult: verifyResult.success ? verifyResult.data : null
        }
      };
    } catch (error) {
      console.error('Recharge credit failed:', error.message);
      return {
        success: false,
        error: error.message
      };
    }
  }

  // Achat de forfait (VOUCHER)
  async purchaseVoucher(phoneNumber, payItemId, customerInfo) {
    try {
      // Pour les forfaits, le montant est souvent inclus dans le payItemId
      // On peut utiliser un montant par défaut ou le récupérer des détails du produit
      const amount = 1000; // Montant par défaut, à ajuster selon le forfait

      // Étape 1: Créer un devis
      const quoteResult = await this.createQuote(payItemId, amount);
      if (!quoteResult.success) {
        return quoteResult;
      }

      const quoteId = quoteResult.data.quoteId;
      const transactionId = `VR${Date.now()}`;

      // Étape 2: Exécuter le paiement
      const customerData = {
        phone: customerInfo.phone,
        email: customerInfo.email,
        name: customerInfo.name,
        address: customerInfo.address,
        serviceNumber: phoneNumber
      };

      const collectResult = await this.collectPayment(quoteId, customerData, transactionId);
      if (!collectResult.success) {
        return collectResult;
      }

      // Étape 3: Vérifier la transaction
      const verifyResult = await this.verifyTransaction(transactionId);

      return {
        success: true,
        data: {
          transactionId: transactionId,
          quoteId: quoteId,
          collectResult: collectResult.data,
          verifyResult: verifyResult.success ? verifyResult.data : null
        }
      };
    } catch (error) {
      console.error('Purchase voucher failed:', error.message);
      return {
        success: false,
        error: error.message
      };
    }
  }

  // Dépôt d'argent (CASHIN)
  async depositMoney(amount, payItemId, customerInfo) {
    try {
      const quoteResult = await this.createQuote(payItemId, amount);
      if (!quoteResult.success) {
        return quoteResult;
      }

      const quoteId = quoteResult.data.quoteId;
      const transactionId = `DP${Date.now()}`;

      const customerData = {
        phone: customerInfo.phone,
        email: customerInfo.email,
        name: customerInfo.name,
        address: customerInfo.address,
        serviceNumber: customerInfo.phone // Pour les dépôts, on utilise le même numéro
      };

      const collectResult = await this.collectPayment(quoteId, customerData, transactionId);
      if (!collectResult.success) {
        return collectResult;
      }

      const verifyResult = await this.verifyTransaction(transactionId);

      return {
        success: true,
        data: {
          transactionId: transactionId,
          quoteId: quoteId,
          collectResult: collectResult.data,
          verifyResult: verifyResult.success ? verifyResult.data : null
        }
      };
    } catch (error) {
      console.error('Deposit money failed:', error.message);
      return {
        success: false,
        error: error.message
      };
    }
  }

  // Retrait d'argent (CASHOUT)
  async withdrawMoney(amount, payItemId, customerInfo) {
    try {
      const quoteResult = await this.createQuote(payItemId, amount);
      if (!quoteResult.success) {
        return quoteResult;
      }

      const quoteId = quoteResult.data.quoteId;
      const transactionId = `WD${Date.now()}`;

      const customerData = {
        phone: customerInfo.phone,
        email: customerInfo.email,
        name: customerInfo.name,
        address: customerInfo.address,
        serviceNumber: customerInfo.payerNumber // Numéro vers lequel retirer
      };

      const collectResult = await this.collectPayment(quoteId, customerData, transactionId);
      if (!collectResult.success) {
        return collectResult;
      }

      const verifyResult = await this.verifyTransaction(transactionId);

      return {
        success: true,
        data: {
          transactionId: transactionId,
          quoteId: quoteId,
          collectResult: collectResult.data,
          verifyResult: verifyResult.success ? verifyResult.data : null
        }
      };
    } catch (error) {
      console.error('Withdraw money failed:', error.message);
      return {
        success: false,
        error: error.message
      };
    }
  }
}

module.exports = MavianceService;